ros/angel_system_nodes/angel_system_nodes/dummy_multi_task_monitor.py

```diff
@@ -0,0 +1,360 @@
+        with open(self._config_file, "r") as f:
+            config = yaml.safe_load(f)
+
+        self._task_state_dict: Dict[int, TaskStateInformation] = {}
+        for t in config["tasks"]:
+            # Load the config for this task
+            with open(t["config_file"], "r") as f:
+                task_config = yaml.safe_load(f)
+
+            n_steps = task_config["steps"][-1]["id"]
+
+            task_state_info = TaskStateInformation(
+                f"{t['id']}_{task_config['title']}",
+                n_steps,
+            )
+
+            # Load the list of task steps
+            task_state_info.steps = [
+                step["description"] for step in task_config["steps"]
+            ]
+
+            self._task_state_dict[t["id"]] = task_state_info
+
+        log.info("Loaded the following tasks:")
+        for key, value in self._task_state_dict.items():
+            log.info(f"    Task {key}: {value.task_title}")
+
+        # Initialize ROS hooks
+        self._task_update_publisher = self.create_publisher(
+            TaskUpdate, self._task_state_topic, 1
+        )
+        self._task_graph_service = self.create_service(
+            QueryTaskGraph, self._query_task_graph_topic, self.query_task_graph_callback
+        )
+
+        # Store module for local use.
+        self._keyboard = keyboard
+        # Start the keyboard monitoring thread
+        log.info("Starting keyboard threads")
+        self._keyboard_t = Thread(target=self.monitor_keypress)
+        self._keyboard_t.daemon = True
+        self._keyboard_t.start()
+        log.info("Starting keyboard threads... done")
+
+    def _clean_skipped_cache(self, current_hmm_step_id: int) -> None:
+        """
+        Clear the error cache any IDs greater than the given.
+        """
+        s = np.asarray(list(self._steps_skipped_cache))
+        to_remove_ids = s[s > current_hmm_step_id]
+        if to_remove_ids.size:
+            self.get_logger().info(
+                f"Clearing step IDs from skipped cache: " f"{to_remove_ids.tolist()}"
+            )
+        [self._steps_skipped_cache.remove(_id) for _id in to_remove_ids]
+
+    def publish_task_state_message(
+        self,
+        task_id: int,
+    ) -> None:
+        """
+        Forms and sends a `angel_msgs/TaskUpdate` message to the
+        TaskUpdates topic.
+
+        :param task_id: Task ID of the task state that is being published.
+        """
+        log = self.get_logger()
+
+        task_state = self._task_state_dict[task_id]
+
+        message = TaskUpdate()
+
+        # Populate message header
+        message.header.stamp = self.get_clock().now().to_msg()
+        message.header.frame_id = "Task message"
+        message.task_name = task_state.task_title
+        if task_state.latest_act_classification_end_time is not None:
+            message.latest_sensor_input_time = nano_to_ros_time(
+                int(task_state.latest_act_classification_end_time * 1e9)
```

Comment:
FWIW the `1e9` is defined as a constant in the same module you imported `nano_to_ros_time` from that is pre-integer-ized.